mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -163,42 +173,73 @@ impl WalletConnectCtx {
-    /// Uses increasing delay between retry attempts starting from 1sec and increase exponentially.
-    /// After successful connection, attempts to restore previous session state from storage.
-    pub(crate) async fn spawn_connection_initialization_fut(
+    /// Centralised task owning **all** connection logic (connect → monitor → reconnect).
+    async fn connection_lifecycle_task(
         self: Arc<Self>,
-        connection_live_rx: UnboundedReceiver<Option<String>>,
+        mut conn_status_rx: UnboundedReceiver<Option<String>>,
+        connection_state_tx: watch::Sender<ConnectionState>,
     ) {
-        info!("Initializing WalletConnect connection");
-        let mut retry_count = 0;
-        let mut retry_secs = 1;
-
-        // Connect to WalletConnect relay client(retry until successful) before proceeeding with other initializations.
-        while let Err(err) = self.connect_client().await {
-            retry_count += 1;
-            error!(
-                "Error during initial connection attempt {}: {:?}. Retrying in {retry_secs} seconds...",
-                retry_count, err
-            );
-            Timer::sleep(retry_secs as f64).await;
-            retry_secs = std::cmp::min(retry_secs * 2, MAX_BACKOFF);
-        }
-
-        // Initialize storage
         if let Err(err) = self.session_manager.storage().init().await {
             error!("Failed to initialize WalletConnect storage, shutting down: {err:?}");
+            let _ = connection_state_tx.send(ConnectionState::Disconnected);
             self.abortable_system.abort_all().error_log();
-        };
+            return;
+        }
 
-        // load session from storage
-        if let Err(err) = self.load_session_from_storage().await {
+        if let Err(err) = self.load_sessions_from_storage().await {
             error!("Failed to load sessions from storage, shutting down: {err:?}");
+            let _ = connection_state_tx.send(ConnectionState::Disconnected);
             self.abortable_system.abort_all().error_log();
+            return;
+        }
+
+        loop {
+            connection_state_tx.send(ConnectionState::Connecting).error_log();
+            info!("WalletConnect: connecting…");
+
+            let mut backoff = 1;
+            while let Err(e) = self.connect_and_subscribe().await {
+                error!("Connection attempt failed: {e:?}; retrying in {backoff}s");
+                Timer::sleep(backoff as f64).await;
+                backoff = std::cmp::min(backoff * 2, MAX_BACKOFF);
+            }
+
+            connection_state_tx.send(ConnectionState::Connected).error_log();
+            info!("WalletConnect: online.");
+
+            if let Some(msg) = conn_status_rx.next().await {
+                info!("WalletConnect: disconnected with message: {msg:?}, will reconnect.");
+            } else {
+                connection_state_tx.send(ConnectionState::Disconnected).error_log();
```

Comment:
don't we wanna abort the system here?

Reply:
Missed it

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2508/commits/e7ef0e1293e2e2811185c0e3fedcc1ae6675d917